Sagging compose box topic w/ emoji at 16px

gusain71 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #31153  <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

All screenshots were captured on the external monitor using Chrome.

**Regular mode:**

| Before | After |
| --- | --- |
| ![before_fix_regular_mode_text](https://github.com/user-attachments/assets/38d38acb-2441-46ca-bb6e-d04a20f84f40) | ![after_fix_regular_mode_text](https://github.com/user-attachments/assets/ec94e7ca-d88e-4768-82b3-2a5b3ad7a622)
| ![before_fix_regular_mode_emoji](https://github.com/user-attachments/assets/dff497e6-0187-4e14-99e8-fa3430dad4c3) | ![after_fix_regular_mode_emoji](https://github.com/user-attachments/assets/97ccbad8-07ed-4711-be0f-26dbba1dea95)


**Compact mode:**

| Before | After |
| --- | --- |
| ![before_fix_compact_mode_text](https://github.com/user-attachments/assets/5fbefd11-56d6-4431-925c-88f4e201f43d) | ![after_fix_compact_mode_text](https://github.com/user-attachments/assets/543ec5d2-eccb-4c17-8e82-08c4b4e55367)
| ![before_fix_compact_mode_emoji](https://github.com/user-attachments/assets/771bb91e-d8ec-412f-bf0a-488b40787257) | ![after_fix_compact_mode_emoji](https://github.com/user-attachments/assets/0ef2b7e7-5c2b-4a5e-a5b8-7009c8436b7f)

I was able to recreate the issue on Safari and Chrome for my external monitor. Specifying the line-height solves this problem. The padding has to be reduced slightly in order to fit the emoji which is being rendered larger on low-resolution screens by the browser in order to retain the detail. The line-height value and the top and bottom padding has to be set using em for the solution to work in both regular and compact mode. 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

gusain71 committed:
compose-box: Fix sagging text on external monitors.

Having an emoji in the topic name pushes down the text
in the compose box topic field. This happens specifically
for low-resolution external monitors on Safari and Chrome.
The emoji is being rendered larger on the low-resolution
monitor to retain the detail by the browser. This makes
the emoji too large to fit within the box during the
current styling. Adding line-height with an em value and
switching the top and bottom padding to an em value fixes
this issue.

Fixes #31153.

alya commented:
Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)? 

You will also need to complete the self-review checklist.

gusain71 commented:
> Thanks for the fix! Could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?
> 
> You will also need to complete the self-review checklist.

Hi @alya !

I have now updated my commit message to match the style guidelines and completed the self-review checklist.

karlstolley commented:
I'm unable to replicate the original sagging look with my low-res testing approach (i.e., opening Firefox in low-res mode on an external monitor). At the same time, I don't detect any regressions here, either.

@alya, perhaps you'd like to try and replicate the original problem on `main` and see if this remedies it?

karlstolley commented:
I went back and reread my message [on the original CZO thread](https://chat.zulip.org/#narrow/stream/9-issues/topic/.F0.9F.93.81.20sagging.20compose.20box.20topic.20w.2F.20emoji.20at.2016px/near/1890099), and I can reproduce in Chrome. While this does appear to correct things at 16/140, in compact mode, even ordinary topics are being raised higher in the compose box:

| Before | After |
| --- | --- |
| ![ordinary-before](https://github.com/user-attachments/assets/7aa4c32b-a89f-4f6c-989b-67955efd05a8) | ![ordinary-after](https://github.com/user-attachments/assets/2e554152-6ff2-4255-8860-abcfa7593b76) |

karlstolley commented:
@gusain71 can you update your screenshots at the very top, and also capture the same sequence in compact mode? Also, if you can add table markup to display these side by side, that would be really useful. Here's the starter code I use:

```txt
| Before | After |
| --- | --- |
|
```

gusain71 commented:
> @gusain71 can you update your screenshots at the very top, and also capture the same sequence in compact mode? Also, if you can add table markup to display these side by side, that would be really useful. Here's the starter code I use:
> 
> ```
> | Before | After |
> | --- | --- |
> |
> ```

Thanks for the starter code! It should be good now. Updated images, description and added markdown table. @karlstolley

karlstolley commented on web/styles/compose.css:
```diff
@@ -931,9 +931,12 @@ textarea.new_message_textarea {
         outline: none;
 
-        padding: 4px 6px;
+        padding: 0.15em 0.43em;
```
Since ems are relative values, it's really important to preserve how they were calculated as comments in the CSS. There's a bunch of examples throughout the codebase where others, including me, have done this.

karlstolley replied:
(Same holds true for the line-height value you specify below.)

gusain71 replied:
Calculations for em values now added to comments.

karlstolley replied:
No, this is not what should be happening here. The codebase should indicate where `0.15em` originated, not what it computes to at different sizes. This is made even more puzzling because where did values like 6.88px come from, compared to the original whole-pixel values on the padding.

gusain71 replied:
Just so I understand this correctly, I should add a comment explaining why we needed to change from 4px to 0.15em for the top and bottom padding? Something like: 

> "Top and bottom padding needs to be a dynamic value to fit emojis in both compact and regular mode on all monitors."

The right and left padding does not actually need to be an em value, keeping it at 6px works fine as it only is the top and bottom padding that is part of the problem. Perhaps this is better and less confusing. 6.88px is just a byproduct of trying to set a value around 6px using em at 14px and 16px.

karlstolley replied:
That's probably fine. Em values should be calculated from pixel values, and if the pixel value you're working with is also an adjustment, that should be noted somewhere too (probably in your commit or the PR).

gusain71 replied:
I have updated the commit with the new changes.

karlstolley commented:
Okay. Not seeing any problems with this in Firefox, including low-res mode. But low-res Chrome is a different story--the alignment of the input text while viewing this PR is not the same as the alignment on the channel dropdown. Screenshot of a guide in an image editor:

![Screenshot 2024-12-16 at 3 02 04 PM](https://github.com/user-attachments/assets/44d894fd-a3d9-4217-8c0a-734b4b91b894)

karlstolley commented:
Closing due to staleness and lack of contributor response.
